### DISCUSSION TITLE
Type mismatch for the first argument of `Tag`

### DISCUSSION THREAD

**Discussion by adebiyial** [markdoc/markdoc#310] [Bug report]
When [customizing Markdoc nodes](https://markdoc.dev/docs/nodes#customizing-markdoc-nodes) for example, the `transform` function is called and it returns a `new Tag`. The first argument of `Tag` is the "html/react" element that should be returned. However, `Tag` always expects a string leading to an error like:

> Argument of type '({ ordered, id, className, children }: IListProps) => Element' is not assignable to parameter of type 'string'.ts(2345)

See code example:

```ts
import { Config, Node, Tag } from '@markdoc/markdoc';
import { List } from '../../components/List';

export default {
  render: List,
  children: ['list'],
  attributes: {
    id: { type: String },
    ordered: { type: Boolean, required: true, default: true },
    className: { type: String },
  },
  transform(node: Node, config: Config) {
    const attributes = node.transformAttributes(config);
    const children = node.transformChildren(config);

    return new Tag(this.render, { ...attributes }, children);
  },
};
```

**Comment by rpaul-stripe** [comment]
In your example, is `List` a React component? You can't actually pass a React component into the render tree. It has to be a string name. When you invoke the React renderer on the client, you pass in a mapping of strings to actual React components. This is because in most usage scenarios the `transform` process happens on the server side completely decoupled from React and the actual React rendering happens on the client side.

```
export default {
  render: "List",
  ...
}
```

And on the client:

```
import { List } from '../../components/List';
Markdoc.renderers.react(content, React, { List });
```

**Reply by adebiyial** [reply]
`List` is a react component that is passed as `this.render` and returned from the component as `JSX.Element`

---

This is an example from the docs itself:

```js
import { Tag } from '@markdoc/markdoc';
import { Heading } from '../../components/Heading';

function generateID(children, attributes) {
  if (attributes.id && typeof attributes.id === 'string') {
    return attributes.id;
  }
  return children
    .filter((child) => typeof child === 'string')
    .join(' ')
    .replace(/[?]/g, '')
    .replace(/\s+/g, '-')
    .toLowerCase();
}

export default {
  render: Heading,
  children: ['inline'],
  attributes: {
    id: { type: String },
    level: { type: Number, required: true, default: 1 },
    className: { type: String }
  },
  transform(node, config) {
    const attributes = node.transformAttributes(config);
    const children = node.transformChildren(config);
    const id = generateID(children, attributes);

    return new Tag(this.render, { ...attributes, id }, children);
  }
};
```

I understand what you explained, but calling `this.render` refers to the react component `Heading`.

**Reply by rpaul-stripe** [reply]
Can you show me where that example is in the documentation? The code example on the page you linked to in the first post [is different](https://markdoc.dev/docs/nodes#customizing-markdoc-nodes:~:text=%60h%24%7Bnode.attributes%5B%27level%27%5D%7D%60).

The Markdoc [Next.js plugin](https://github.com/markdoc/next.js) allows using a React component as the value of the `render` attribute, but it has some [custom typing](https://github.com/markdoc/next.js/blob/main/src/index.d.ts#L16-L19) and implementation details that enable that behavior in a way that is specific to Next.js environments. I wonder if maybe the example you saw was for the Next.js plugin, or maybe there's a place where we incorrectly show a Next.js example in the regular documentation.

**Reply by adebiyial** [reply]
Here's a [line](https://github.com/markdoc/docs/blob/main/markdoc/nodes/heading.markdoc.js#L29) from the Markdoc documentation repo.

I'm using Next.js. I think your concluding answer might be what I need.

**Reply by segphault** [reply]
Oh yeah, that is in the Markdoc documentation site, which is powered by Next.js. If you are using Next.js and have the `@markdoc/next.js` npm package installed in your project's `node_modules` folder, you shouldn't be seeing the error message.

**Reply by adebiyial** [reply]
Yeah, I have everything in place. But the error is there.

<img width="1568" alt="image" src="https://user-images.githubusercontent.com/18689106/210841012-2f2b17e4-0a0d-4446-8b0e-177ca2e5eac1.png">

PS: VSCode shows the error when I view the Markdoc docs repo itself.
